### PR TITLE
Add option in data loader for out of order data

### LINKED ISSUE #105203: Pytorch dataloader not loading first-available data with multiple workers
### 🐛 Describe the bug

When using a dataloader with num_workers > 1 the batches are constructed in parallel to speed up the data-loading. I would expect that the dataloader returns the first-available data (FIFO-queue) to make sure that it runs as fast as possible. However it seems that each process will take turns returning data which slows don't data-loading quite significantly.

Below is a minimal code example:

```python
import torch
import math
import time

class MyIterableDataset(torch.utils.data.IterableDataset):
    def __init__(self, start, end):
        super(MyIterableDataset).__init__()
        assert end > start, "this example code only works with end >= start"
        self.start = start
        self.end = end

    def give_data(self, start, end):
        for i in range(start, end):
            if i > 10:
                time.sleep(2)
            yield i

    def __iter__(self):
        worker_info = torch.utils.data.get_worker_info()
        if worker_info is None:  # single-process data loading, return the full iterator
            iter_start = self.start
            iter_end = self.end
        else:  # in a worker process
            # split workload
            per_worker = int(math.ceil((self.end - self.start) / float(worker_info.num_workers)))
            worker_id = worker_info.id
            iter_start = self.start + worker_id * per_worker
            iter_end = min(iter_start + per_worker, self.end)
        return self.give_data(iter_start, iter_end)
    
if __name__ == "__main__":
    ds = MyIterableDataset(start=0, end=20)

    # Mult-process loading with two worker processes
    for item in (torch.utils.data.DataLoader(ds, num_workers=2, batch_size=2)):
        print(item)
```

The result of this script is:
```python
tensor([0, 1]) # Loaded fast
tensor([10, 11])  # Loaded slowly
tensor([2, 3])  # Loaded fast
tensor([12, 13])  # Loaded slowly
tensor([4, 5])  # Loaded fast
tensor([14, 15])  # Loaded slowly
tensor([6, 7])  # Loaded fast
tensor([16, 17])  # Loaded slowly
tensor([8, 9])  # Loaded fast
tensor([18, 19])  # Loaded slowly
```

However I would expect something like to be the result:
```python
tensor([0, 1]) # Loaded fast
tensor([2, 3])  # Loaded fast
tensor([4, 5])  # Loaded fast
tensor([6, 7])  # Loaded fast
tensor([10, 11])  # Loaded slowly
tensor([8, 9])  # Loaded fast
tensor([12, 13])  # Loaded slowly
tensor([14, 15])  # Loaded slowly
tensor([16, 17])  # Loaded slowly
tensor([18, 19])  # Loaded slowly
```

### Versions

Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.24.1
Libc version: N/A

Python version: 3.11.0 (main, Nov 30 2022, 13:48:51) [Clang 13.1.6 (clang-1316.0.21.2.5)] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy==1.3.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.0
[pip3] pytorch-lightning==2.0.3
[pip3] torch==2.0.1
[pip3] torchmetrics==0.11.4
[conda] Could not collect

cc @SsnL @VitalyFedyunin @ejguan @dzhulgakov

### PR DESCRIPTION
**Pull request by michael-diggin**
Fixes #105203

Facing a similar problem to the linked issue, where variable sized input data can mean that a handful of slow to process samples holds up smaller and faster to process samples from being used. This also leads to lower GPU utilization as well. In certain cases, e.g. evaluation epochs, inference pipelines or other cases where reproducibility isn't important, this can bring significant speed ups.

This PR adds an `allow_out_of_order` bool input to the `DataLoader` class, defaulting to `false`, which when set to `true` will returning data from workers in whatever order they are ready/processed in, rather in the strict index order.
Instead of storing data that was returned out of order, it is passed directly to the main thread and the entry in `_task_info` is deleted. The main changes are they to check that an entry in `_task_info` does exist, and only increasing `self._rcvd_idx` when the lowest index remaining gets returned.

Two tests are added to test this for iterable type datasets and index type datasets.

cc @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov

### CODE REVIEW

**Review by andrewkho — CHANGES_REQUESTED** [review]
Thanks for adding this feature! After a first pass it looks fine to me, just a couple of small changes/suggestions

File: torch/utils/data/dataloader.py
Comment (andrewkho) [inline torch/utils/data/dataloader.py:218]: Another warning to call out: in the case of imbalanced data it may lead to skewed data distribution being fed to the trainer, where eg batches with small data/faster-to-pre-proc may introduce unwanted data-skew.

File: torch/utils/data/dataloader.py
Comment (michael-diggin) [reply torch/utils/data/dataloader.py]: That's nicer, thanks!

**Comment by michael-diggin** [comment]
> Thanks for adding this feature! After a first pass it looks fine to me, just a couple of small changes/suggestions

Thanks for the quick review @andrewkho! I've made the changes you've suggested.

File: torch/utils/data/dataloader.py
Comment (andrewkho) [inline torch/utils/data/dataloader.py:218]: One potential issue here that occurred to me: since work is continuously being distributed to workers in round-robin fashion, we can hit a scenario where one slow or blocked worker can end up holding all the tasks, blocking other tasks from being scheduled.

As a follow up, when in_order is False, we should also distribute work to workers to keep them balanced since reproducibility is already given up in these scenarios.

Comment (michael-diggin) [reply torch/utils/data/dataloader.py:218]: Yep! That issue occurs currently with in_order=True I think (out of order tasks are buffered up instead of being given out)?
Would like this added as a warning, and fixed in a later PR, or fixed as part of this PR?
My _guess_ would be to track the number of tasks given to each worker and check [here](https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L1476) if it's < the prefetch factor.

Comment (andrewkho) [reply torch/utils/data/dataloader.py:218]: I think it'd be fine to do in a separate PR, given this is default-off, although I think we'd like it to land together in the next release

Comment (michael-diggin) [reply torch/utils/data/dataloader.py:218]: Yep I'm very happy to do the follow up PR!

**Review by andrewkho — APPROVED** [review]

**Comment by andrewkho** [comment]
Please wait for CI to pass and then land

**Comment by michael-diggin** [comment]
Ah sorry @andrewkho, I completely forgot that running merge with `-r` would require workflows that need your approval.

**Comment by michael-diggin** [comment]
Hi @andrewkho - small bump on this. Would you be able to approve the workflows/CI?
There’s not been any changes since you approved the PR before, just an accidental rebase that requires rerunning the workflows. Thanks!

**Comment by andrewkho** [comment]
@michael-diggin  sorry for delay, just kicked off the workflows, thanks for your patience!

**Pull request closed without merging**
